File: src/main/java/net/serebryansky/nology/controller/TripController.java
package net.serebryansky.nology.controller;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import net.serebryansky.common.service.PlaceService;
import net.serebryansky.common.service.RoutingService;
import net.serebryansky.nology.model.TripGenerationRequest;
import net.serebryansky.nology.model.Trip;
import net.serebryansky.common.service.CityService;
import net.serebryansky.common.service.TripBuilderService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import static net.serebryansky.common.util.LoggingUtil.logOnEach;
import static org.springframework.http.ResponseEntity.ok;

@RestController
@RequestMapping("trips")
@Slf4j
@RequiredArgsConstructor
public class TripController {

    private final TripBuilderService tripBuilderService;
    private final CityService cityService;
    private final PlaceService placeService;
    private final RoutingService routingService;

    @PostMapping("generate")
    public ResponseEntity<Mono<Trip>> generate(@RequestBody TripGenerationRequest request) {
        val trip = new Trip();
        val city = cityService.getCity(request.getCityId())
                .cache();
        val builderResponse = tripBuilderService.buildTrip(trip)
                .cache();
        Flux<String> placeIds = builderResponse
                .flux()
                .flatMap(response -> Flux.fromIterable(response.getPlaceIds()));
        val places = placeIds
                .flatMap(placeService::getPlace)
                .cache();
        val routes = places
                // send requests to routing
                .buffer(2, 1)
                .filter(p -> p.size() == 2)
                .flatMap(routingService::getRoute)
                .cache();
        val result = Mono.empty()
                .doOnEach(logOnEach(() -> log.info("Generate {}", request)))
                // send request to city
                .then(city)
                .doOnNext(trip::setCity)
                // send request to builder
                .then(builderResponse)
                // send requests to place service
                .thenMany(places)
                .collectList()
                .doOnNext(trip::setPlaces)
                // send requests to routing
                .thenMany(routes)
                .collectList()
                .doOnNext(trip::setRoutes)
                .then(Mono.just(trip));
        return ok(result);
    }








}

File: src/main/kotlin/net/serebryansky/nology/controller/TripController.kt
package net.serebryansky.nology.controller

import kotlinx.coroutines.reactive.awaitFirst
import mu.KotlinLogging
import net.serebryansky.common.service.CityService
import net.serebryansky.common.service.PlaceService
import net.serebryansky.common.service.RoutingService
import net.serebryansky.common.service.TripBuilderService
import net.serebryansky.nology.model.Trip
import net.serebryansky.nology.model.TripGenerationRequest
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("trips")
class TripController(
    private val tripBuilderService: TripBuilderService,
    private val cityService: CityService,
    private val placeService: PlaceService,
    private val routingService: RoutingService
) {

    private val log = KotlinLogging.logger {}

    @PostMapping("generate")
    suspend fun generate(@RequestBody request: TripGenerationRequest): ResponseEntity<Trip> {
        log.info { "Generate $request" }
        val trip = Trip()
        val city = cityService.getCity(request.cityId)
        val builderResponse = tripBuilderService.buildTrip(trip)
        val placeIds = builderResponse.placeIds
        val places = placeIds.map { placeService.getPlace(it) }
        val routes = places // send requests to routing
            .windowed(2)
            .filter { it.size == 2 }
            .map { routingService.getRoute(it) }
        trip.city = city
        trip.places = places
        trip.routes = routes
        return ResponseEntity.ok(trip)
    }
}
